pytest_rerunfailures.py

```diff
@@ -71,151 +100,241 @@ def check_options(config):
+            self._execute_reruns()
 
+        return True
 
-def pytest_runtest_protocol(item, nextitem):
-    """
-    Note: when teardown fails, two reports are generated for the case, one for
-    the test case and the other for the teardown error.
-    """
+    def _execute_reruns(self):
+        """
+        Perform reruns for failed tests
+        """
+        for item in self.tests_to_rerun:
+            reruns = self._get_reruns_count(item)
+            if reruns is None:
+                continue
 
-    reruns = get_reruns_count(item)
-    if reruns is None:
-        # global setting is not specified, and this test is not marked with
-        # flaky
-        return
-
-    # while this doesn't need to be run with every item, it will fail on the
-    # first item if necessary
-    check_options(item.session.config)
-    delay = get_reruns_delay(item)
-    parallel = hasattr(item.config, 'slaveinput')
-
-    for i in range(reruns + 1):  # ensure at least one run of each item
-        item.ihook.pytest_runtest_logstart(nodeid=item.nodeid,
-                                           location=item.location)
-        reports = runtestprotocol(item, nextitem=nextitem, log=False)
+            self._rerun_item(item, reruns)
 
-        for report in reports:  # 3 reports: setup, test, teardown
-            report.rerun = i
-            xfail = hasattr(report, 'wasxfail')
-            if i == reruns or not report.failed or xfail:
-                # last run or no failure detected, log normally
-                item.ihook.pytest_runtest_logreport(report=report)
-            else:
-                # failure detected and reruns not exhausted, since i < reruns
-
-                # collect all item related fixtures and call finalizers for them
-                fixturemanager = item.session._fixturemanager
-                fixtures = set(item.fixturenames)
-                fixtures.update(fixturemanager._getautousenames(item.nodeid))
-                fixtures.update(item._fixtureinfo.argnames)
-                usefixtures = getattr(item.function, 'usefixtures', None)
-                if usefixtures:
-                    fixtures.update(usefixtures.args)
-
-                for fixt in fixtures:
-                    for fixtdef in fixturemanager.getfixturedefs(fixt, item.nodeid) or []:
-                        item._initrequest()
-                        fixtdef.finish(item._request)
-
-                # failure detected and reruns not exhausted, since i < reruns
-                report.outcome = 'rerun'
-                time.sleep(delay)
 
+    def _rerun_item(self, item, reruns):
+        """
+        Perform reruns for single test items
+
+        Parameters
+        ----------
+        item : _pytest.main.Item
+        """
+        delay = self._get_reruns_delay(item)
+        parallel = hasattr(item.config, 'slaveinput')
+
+        for i in range(reruns):
+            self._invalidate_fixtures(item)
+            time.sleep(delay)
```

Comment:
why do we need this?

Reply:
It is adaptation of base rerunfailures plugin - ability to rerun tests with some delay. It is controlled by `--reruns-delay` param